kroxylicious-operator/src/main/java/io/kroxylicious/operator/ProxyConfigSecret.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.kroxylicious.operator;
```

Comment:
Should we be encoding an API version in the package here as well? So that it is clear that this is the operator code for reconciling a particular API version e.g

`io.kroxylicious.operator.v1alpha(1)?.ProxyConfigSecret`
`io.kroxylicious.operator.v1beta(1)?.ProxyConfigSecret` ?

Reply:
The JOSDK docs don't really cover multiple versions, but there has been [a question](https://github.com/operator-framework/java-operator-sdk/discussions/1710) and they have tests for it. I think it shakes out like this:

* With a schema-first approach, if you declare multiple versions in the CR then you'll get multiple classes generated. The [fabric8 generator](https://github.com/fabric8io/kubernetes-client/blob/main/doc/java-generation-from-CRD.md) has limited control over the [naming the generated types](https://github.com/fabric8io/kubernetes-client/blob/main/java-generator/core/src/main/java/io/fabric8/java/generator/CRGeneratorRunner.java). For example, you cannot encode the version in type names, only in package names. There is also support for naming some types as being `-existing-java-types`, which will prevent generation, meaning you can, I suppose, hand-maintain some of the types if you wanted to.
* Within the JOSDK API, a `Reconciler` has a type argument for the CR class, which then references the spec and the status types, and so on all the way down. ⇒ A Reconciler is a per-version thing. Likewise for a `DependentResource`. So if you have parallel schema types for different versions then you're forced into the pattern of using parallel operator types too. Obviously its up to you to decide how these are named (e.g. encoding the version into the type or package name).

It is unappealing to have to duplicate all your operator logic just because you added a new CR version if "most of" the Open API schema was be the same between the two CR versions. You'd want "most of" the operator Java code _ought_ to be the same too. 

* One way to handle this would be to apply the "abstract method" pattern to `DependentResource`, providing  methods in subclasses to handle the particular concrete types in the generated Java API, but keeping most of the logic the same. This would probably require the use of type parameters in the abstract class. It could get messy if there were a larger number of differences between the APIs. But in that case you should probably give up on trying to treat the two APIs "the same" and admit they're different.
* An alternative would be to rely on Kube's behaviour wrt serving resources at different versions. Kube's model assumes that the API server can convert between schema versions in order to serve a version understood by the client. It can do that either because there are no syntactic differences (i.e. the `NONE` conversion), or because the CRD has declared the use of Webhook for conversion so the apiserver can call out to get a CR converted. If the operator author is willing to rely on the CR version serving as a guarantee, then they could just generate the Java API for the latest CRD schema version (by presenting a different CRD with a single version in to the Fabric8 generator, for example), and write the operator (for different versions) using a single set of classes. We might need to maintain per-version root classes (i.e. the CR class) by hand, but that's not really much code.
    * Of course if we can't rely on the NONE conversion, this just moves the problem to the webhook, which we'd also have to write. But that might be needed anyway (so that our API can be converted as Kube expects), and it also would probably be less code. There would also be the option of operating using untyped `JsonNode` trees I suppose.

So all this overthinking is to say the answer to your question is "Maybe, it depends" (on how our schema evolves). The answer to this question for the _operator_ code is not something that will be written in stone and used against us in the future because it's not public API. So we can figure out our answer when we're actually ready to support multiple API versions. 

It's a little less obvious for the Java API reflecting the CR API. That feels like a more consequential decision (because someone might want to use those classes as a public API), and its a decision which we need to make now. Let's remember that our own system tests will be one such client of this API.

We can look at the approach that Fabric8 has taken for the `kubernetes-model` artifacts. With the exception of `kubernetes-model-core` and a handful of other artifacts, most of this Java APIs are encoding the resource version in the package name. If we were to follow the same direction it means we'd need to be duplicating the system test code, or needing to work with some abstractions over different versions in the tests too.